wrapper/src/main/java/software/amazon/jdbc/plugin/strategy/fastestresponse/FastestResponseStrategyPlugin.java

```diff
@@ -0,0 +1,199 @@
+  }
+
+  @Override
+  public Set<String> getSubscribedMethods() {
+    return subscribedMethods;
+  }
+
+  @Override
+  public Connection connect(
+      final String driverProtocol,
+      final HostSpec hostSpec,
+      final Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> connectFunc)
+      throws SQLException {
+
+    Connection conn = connectFunc.call();
+    if (isInitialConnection) {
+      this.hostResponseTimeService.setHosts(this.pluginService.getHosts());
+    }
+    return conn;
+  }
+
+  @Override
+  public Connection forceConnect(
+      final String driverProtocol,
+      final HostSpec hostSpec,
+      final Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> forceConnectFunc)
+      throws SQLException {
+
+    Connection conn = forceConnectFunc.call();
+    if (isInitialConnection) {
+      this.hostResponseTimeService.setHosts(this.pluginService.getHosts());
+    }
+    return conn;
+  }
+
+  @Override
+  public boolean acceptsStrategy(HostRole role, String strategy) {
+    return FASTEST_RESPONSE_STRATEGY_NAME.equalsIgnoreCase(strategy);
+  }
+
+  @Override
+  public HostSpec getHostSpecByStrategy(final HostRole role, final String strategy)
+      throws SQLException, UnsupportedOperationException {
+
+    if (!acceptsStrategy(role, strategy)) {
+      return null;
+    }
+
+    final HostSpec fastestResponseHost = latestResponseHost.get(role.name());
+    if (fastestResponseHost != null) {
+      HostSpec foundHostSpec = this.pluginService.getHosts().stream()
+          .filter(x -> x.equals(fastestResponseHost))
+          .findAny()
+          .orElse(null);
+
+      if (foundHostSpec != null) {
+        return foundHostSpec;
+      }
+    }
+
+    // Cached result isn't available. Need to find the fastest response time host.
+
+    final HostSpec calculatedFastestResponseHost = this.pluginService.getHosts().stream()
+        .filter(x -> role.equals(x.getRole()))
+        .map(x -> new Tuple(x, this.hostResponseTimeService.getResponseTime(x)))
+        .sorted(Comparator.comparingInt(x -> x.responseTime))
+        .map(x -> x.hostSpec)
+        .findFirst()
+        .orElse(null);
+
+    if (calculatedFastestResponseHost == null) {
+      // Unable to identify the fastest response host.
+      // As a last resort, let's use a random host selector.
+      return randomHostSelector.getHost(this.hosts, role, properties);
+    }
+
+    latestResponseHost.put(role.name(), calculatedFastestResponseHost, this.cacheExpirationNano);
+
+    return calculatedFastestResponseHost;
+  }
+
+  @Override
+  public void notifyNodeListChanged(final Map<String, EnumSet<NodeChangeOptions>> changes) {
+    this.hosts = this.pluginService.getHosts();
+    this.hostResponseTimeService.setHosts(this.hosts);
+  }
+
+  private static class Tuple {
```

Comment:
Tuple is a very generic computing term

if we need this class only for those types, let's call it ResponseTimeTuple